src/algebraic_geometry/Gamma_Spec_adjunction.lean

```diff
@@ -0,0 +1,313 @@
+/-
+Copyright (c) 2021 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import algebraic_geometry.Spec
+import algebraic_geometry.ringed_space
+import topology.sheaves.sheaf_condition.basis_le
+import topology.sheaves.functors
+
+/-!
+# Adjunction between `Γ` and `Spec`
+
+Define the adjunction `Γ_Spec_adjunction : Γ ⊣ Spec` (or more technically,
+`Γ.right_op ⊣ Spec.to_LocallyRingedSpace`) by defining the unit (done in Spec.lean) and
+counit (`to_Γ_Spec`, in multiple steps in this file) and checking that they satisfy
+the left and right triangle identities. The constructions and proofs make use of
+maps and lemmas defined and proved in structure_sheaf.lean extensively.
+
+Notice that since the adjunction is between contravariant functors, you get to choose
+one of the two categories to have arrows reversed, and it is equally valid to present
+the adjunction as `Spec ⊣ Γ` (`Spec.to_LocallyRingedSpace.right_op ⊣ Γ`), in which
+case the unit and the counit would switch to each other.
+-/
+
+noncomputable theory
+universe variables u v
+
+open prime_spectrum
+
+namespace algebraic_geometry
+open opposite
+open category_theory
+open structure_sheaf
+open topological_space
+open algebraic_geometry.LocallyRingedSpace
+open Top.presheaf
+open Top.presheaf.sheaf_condition
+
+
+variable (R : CommRing)
+
+/-- `Spec` as a function from CommRing to LocallyRingedSpace. -/
+abbreviation Spec' := Spec.to_LocallyRingedSpace.obj (op R)
```

Comment:
I would advise against this abbreviation. It makes things shorter, but at the expense of readability. The advantage of the `Spec.to_xxx` is that you directly see which category you land in.

Reply:
I think it's more readable as it's close to the mathematical notation, and it's unlikely to get confused about the category as most of the file is under the LocallyRingedSpace namespace. It's used 11 times so not that convenient to remove, but maybe I should make it private. Similarly `Γ'` is used 27 times (reduces to 15 times if (r : Γ' X) is included in the `variables`).

Reply:
But oftentimes when you use the `Spec'` abbreviation, you are not really working with the category of locally ringed spaces, as there are some coercions going on. For instance, in line 47, instead of `(Spec' R).to_Top`, you can write `Spec.to_Top`. In line 68, all you're really using is the prime spectrum as a *type*, so you can replace `Spec'` with `prime_spectrum` here. Also, in line 91, you can write `X.to_Top ⟶ Spec.to_Top.obj ...`. Similarly, in line 138 you can write `(structure_sheaf (Γ' X)).1` instead of `(Spec' (Γ' X)).presheaf` etc.  The only time when you're really using `Spec'` in the category of ringed spaces is when you state the triangle identities. These can probably be avoided by inlining those proofs into the definition of `Γ_Spec_core_unit_counit`. I'm not saying they should be inlined necessarily, but it is an option.

Reply:
Hmm, in fact my `Spec'` is just `Spec.LocallyRingedSpace_obj`, no need of `to_` and `op` ! And there are also `Spec.Top_obj` etc. These makes things shorter and nicer.

I won't be able to inline the proof of triangle identities because currently `def core_unit_counit` already takes tens of seconds to check.

Reply:
removed Spec' and replaced them according to your suggestion (except I used the shorter versions) at https://github.com/leanprover-community/mathlib/pull/9802/commits/ca95f68380ca78a0ddbe091c98e1ba18dc8277f7